File: paiguangguang/backend/app/chatbot/schemas/stream.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict


ChatStreamStatus = Literal["pending", "streaming", "completed", "failed", "cancelled"]
ChatStreamFinalStatus = Literal["completed", "failed", "cancelled"]
ChatStreamEventName = Literal[
    "message.created",
    "message.delta",
    "message.completed",
    "message.failed",
    "message.cancelled",
    "usage.updated",
    "stream.end",
]


class StreamMessageData(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    sequence_number: int
    status: ChatStreamStatus
    content: str
    model: str | None = None
    updated_at: datetime | None = None


class StreamMessageCreatedData(BaseModel):
    schema_version: Literal["1"] = "1"
    request_id: UUID
    conversation_id: UUID
    assistant_message_id: UUID
    sequence: int
    created_at: datetime
    replayed: bool
    user_message: StreamMessageData
    assistant_message: StreamMessageData


class StreamMessageDeltaData(BaseModel):
    schema_version: Literal["1"] = "1"
    request_id: UUID
    conversation_id: UUID
    assistant_message_id: UUID
    sequence: int
    created_at: datetime
    delta: str
    content_length: int


class StreamCompletedMessageData(BaseModel):
    id: UUID
    status: Literal["completed"] = "completed"
    content: str
    sequence_number: int
    model: str | None = None
    updated_at: datetime


class StreamMessageCompletedData(BaseModel):
    schema_version: Literal["1"] = "1"
    request_id: UUID
    conversation_id: UUID
    assistant_message_id: UUID
    sequence: int
    created_at: datetime
    message: StreamCompletedMessageData
    finish_reason: str | None = None


class StreamErrorData(BaseModel):
    code: str
    message: str
    retryable: bool


class StreamMessageFailedData(BaseModel):
    schema_version: Literal["1"] = "1"
    request_id: UUID
    conversation_id: UUID
    assistant_message_id: UUID
    sequence: int
    created_at: datetime
    error: StreamErrorData
    partial: bool
    content: str


class StreamMessageCancelledData(BaseModel):
    schema_version: Literal["1"] = "1"
    request_id: UUID
    conversation_id: UUID
    assistant_message_id: UUID
    sequence: int
    created_at: datetime
    reason: Literal["user_requested"] = "user_requested"
    content: str


class StreamUsageUpdatedData(BaseModel):
    schema_version: Literal["1"] = "1"
    request_id: UUID
    conversation_id: UUID
    assistant_message_id: UUID
    sequence: int
    created_at: datetime
    prompt_tokens: int | None = None
    completion_tokens: int | None = None
    total_tokens: int | None = None
    source: Literal["provider", "unknown"]


class StreamEndData(BaseModel):
    schema_version: Literal["1"] = "1"
    request_id: UUID
    conversation_id: UUID
    assistant_message_id: UUID
    sequence: int
    created_at: datetime
    final_status: ChatStreamFinalStatus


@dataclass(frozen=True, slots=True)
class ChatStreamEvent:
    event: ChatStreamEventName
    data: BaseModel
    sequence: int


def encode_chatbot_sse_event(event: ChatStreamEvent) -> str:
    import json

    payload = json.dumps(event.data.model_dump(mode="json"), ensure_ascii=False, separators=(",", ":"))
    return f"id: {event.sequence}\nevent: {event.event}\ndata: {payload}\n\n"


def encode_chatbot_sse_keepalive() -> str:
    return ": keepalive\n\n"

File: paiguangguang/frontend/features/chatbot/types/stream.ts
import type { MessageData, MessageStatus } from "./message";

export type StreamSchemaVersion = "1";

export type StreamEventName =
  | "message.created"
  | "message.delta"
  | "message.completed"
  | "message.failed"
  | "message.cancelled"
  | "usage.updated"
  | "stream.end";

export type StreamMessageSnapshot = Pick<
  MessageData,
  "id" | "sequence_number" | "status" | "content" | "model" | "updated_at"
> & {
  updated_at: string | null;
};

export type StreamMessageCreatedData = {
  schema_version: StreamSchemaVersion;
  request_id: string;
  conversation_id: string;
  assistant_message_id: string;
  sequence: number;
  created_at: string;
  replayed: boolean;
  user_message: StreamMessageSnapshot;
  assistant_message: StreamMessageSnapshot;
};

export type StreamMessageDeltaData = {
  schema_version: StreamSchemaVersion;
  request_id: string;
  conversation_id: string;
  assistant_message_id: string;
  sequence: number;
  created_at: string;
  delta: string;
  content_length: number;
};

export type StreamMessageCompletedData = {
  schema_version: StreamSchemaVersion;
  request_id: string;
  conversation_id: string;
  assistant_message_id: string;
  sequence: number;
  created_at: string;
  message: {
    id: string;
    status: "completed";
    content: string;
    sequence_number: number;
    model: string | null;
    updated_at: string;
  };
  finish_reason: string | null;
};

export type StreamErrorData = {
  code: string;
  message: string;
  retryable: boolean;
};

export type StreamMessageFailedData = {
  schema_version: StreamSchemaVersion;
  request_id: string;
  conversation_id: string;
  assistant_message_id: string;
  sequence: number;
  created_at: string;
  error: StreamErrorData;
  partial: boolean;
  content: string;
};

export type StreamMessageCancelledData = {
  schema_version: StreamSchemaVersion;
  request_id: string;
  conversation_id: string;
  assistant_message_id: string;
  sequence: number;
  created_at: string;
  reason: "user_requested";
  content: string;
};

export type StreamUsageUpdatedData = {
  schema_version: StreamSchemaVersion;
  request_id: string;
  conversation_id: string;
  assistant_message_id: string;
  sequence: number;
  created_at: string;
  prompt_tokens: number | null;
  completion_tokens: number | null;
  total_tokens: number | null;
  source: "provider" | "unknown";
};

export type StreamEndData = {
  schema_version: StreamSchemaVersion;
  request_id: string;
  conversation_id: string;
  assistant_message_id: string;
  sequence: number;
  created_at: string;
  final_status: Extract<MessageStatus, "completed" | "failed" | "cancelled">;
};

export type ChatStreamEventPayload =
  | { event: "message.created"; data: StreamMessageCreatedData; sequence: number }
  | { event: "message.delta"; data: StreamMessageDeltaData; sequence: number }
  | { event: "message.completed"; data: StreamMessageCompletedData; sequence: number }
  | { event: "message.failed"; data: StreamMessageFailedData; sequence: number }
  | { event: "message.cancelled"; data: StreamMessageCancelledData; sequence: number }
  | { event: "usage.updated"; data: StreamUsageUpdatedData; sequence: number }
  | { event: "stream.end"; data: StreamEndData; sequence: number };

export type UnknownChatStreamEventPayload = {
  event: string;
  data: unknown;
  sequence: number;
};

export type ParsedChatStreamEvent = ChatStreamEventPayload | UnknownChatStreamEventPayload;

